src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -80,6 +81,110 @@ import java.util.*
 
 
 
+data class Transaction (
+    val transactionDate: LocalDateTime,
+    val transactionType: Type,
+    val amount : Double,
+    val oldBalance : Double,
+    val newBalance : Double,
+    val transactionStatus : Status
+) {
+    enum class Type {
+        DEPOSIT , WITHDRAWAL
+    }
+
+    enum class Status {
+        SUCCESS, FAILURE
+    }
+}
+
+class TransactionalBankAccount(
+    accountNumber: String,
+    accountHolderName: String,
+    private val transactions: MutableList<Transaction> = mutableListOf()
+) : BankAccount(accountNumber, accountHolderName) {
+
+    override fun deposit(amount: Double) {
+        val oldBalance = getBalance
+        super.deposit(amount)
+        transactions.add(Transaction(currentTime, Transaction.Type.DEPOSIT, amount, oldBalance, getBalance, Transaction.Status.SUCCESS))
+    }
+
+    override fun withdraw(amount: Double): Boolean {
+        val oldBalance = getBalance
+        val result = super.withdraw(amount)
+        val status = if(result) Transaction.Status.SUCCESS else Transaction.Status.FAILURE
+
+        transactions.add(Transaction(currentTime, Transaction.Type.WITHDRAWAL, amount, oldBalance, getBalance, status))
+        return result
+    }
+
+    fun getAllTransactions() : List<Transaction> {
+        return transactions.sortedByDescending { it.transactionDate }
```

Comment:
Well done! We should always return a copy of the data instead of the original reference of the transactions list.